### PR TITLE
ENH: Rykoff model of the magnitude uncertainty

### LINKED ISSUE #517: ENH: Rykoff apparent magnitude uncertainty of galaxies
## Description
Add a function that returns the uncertainty of the galaxies' apparent magnitude in the given filter bands as described by equation (6) in Rykoff et al. (2015). Note: the normalisation constant k is absorbed in the included in parameter t_eff

## Inputs
 - Apparent magnitudes
 - 10 sigma apparent magnitude limits
 - Parameters a and b of the model
 - Uncertainty limit

## Outputs
 - Apparent magnitude uncertainty of galaxies

## References
- Rykoff et al. (2015), Assessing galaxy limiting magnitudes in large optical surveys, [arXiv 1509.00870](https://arxiv.org/abs/1509.00870)

### PR DESCRIPTION
**Pull request by philipp128**
## Description
This PR adds a feature to calculate the magnitude uncertainty following the model of Rykoff et al. (2015). Closes #517.
## Checklist
- [x] Follow the [Contributor Guidelines](https://github.com/skypyproject/skypy/blob/main/CONTRIBUTING.rst)
- [x] Write unit tests
- [x] Write documentation strings
- [x] Assign someone from your working team to review this pull request
- [x] Assign someone from the infrastructure team to review this pull request

### CODE REVIEW

**Review by rrjbca — COMMENTED** [review]
I think this should go in skypy.utils.photometry. I looked over the paper quickly and don't think anything in the model is specific to galaxies. This would save having to create an additional super-specific submodule.

File: skypy/galaxies/magnitude_error.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+    flux = 10 ** (-0.4 * (np.subtract(magnitude, magnitude_zp)))
+    flux_limit = 10 ** (-0.4 * (np.subtract(magnitude_limit, magnitude_zp)))
+    t_eff = np.exp(a + b * (np.subtract(magnitude_limit, 21.0)))
+    flux_noise = flux_limit ** 2 * t_eff / 10 ** 2 - flux_limit
+    error = 2.5 / np.log(10) * (1 / (flux * t_eff) * (1 + flux_noise / flux)) ** 0.5
+
+    # replace the values larger than the error limit with the error_limit
+    if error_limit is not None:
+        error = np.where(error > error_limit, error_limit, error)
```
Comment (rrjbca) [inline skypy/galaxies/magnitude_error.py:96]: [`numpy.minimum`](https://numpy.org/doc/stable/reference/generated/numpy.minimum.html)

File: skypy/galaxies/magnitude_error.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    where :math:`a` and :math:`b` are free parameters.
+
+    References
+    ----------
+    .. [1] Rykoff E. S., Rozo E., Keisler R., 2015, eprint arXiv:1509.00870
+
+    """
+
+    flux = 10 ** (-0.4 * (np.subtract(magnitude, magnitude_zp)))
+    flux_limit = 10 ** (-0.4 * (np.subtract(magnitude_limit, magnitude_zp)))
```
Comment (rrjbca) [inline skypy/galaxies/magnitude_error.py:89]: [`skypy.utils.photometry.luminosity_from_absolute_magnitude`](https://skypy.readthedocs.io/en/stable/api/skypy.utils.photometry.luminosity_from_absolute_magnitude.html)

File: skypy/galaxies/magnitude_error.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    t_eff = np.exp(a + b * (np.subtract(magnitude_limit, 21.0)))
+    flux_noise = flux_limit ** 2 * t_eff / 10 ** 2 - flux_limit
```
Comment (rrjbca) [inline skypy/galaxies/magnitude_error.py:91]: [`numpy.square`](https://numpy.org/doc/stable/reference/generated/numpy.square.html)

File: skypy/galaxies/magnitude_error.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    error = 2.5 / np.log(10) * (1 / (flux * t_eff) * (1 + flux_noise / flux)) ** 0.5
```
Comment (rrjbca) [inline skypy/galaxies/magnitude_error.py:92]: [`numpy.sqrt`](https://numpy.org/doc/stable/reference/generated/numpy.sqrt.html)

**Comment by philipp128** [comment]
> I think this should go in skypy.utils.photometry. I looked over the paper quickly and don't think anything in the model is specific to galaxies. This would save having to create an additional super-specific submodule.

@rrjbca hmm yeah. I'll change it. I initially added it to the galaxy module as Rykoff et al. explicitly mention galaxies while deriving the equation. However, it should not be dependent on the object type but solely survey dependent.

File: skypy/galaxies/magnitude_error.py (hunk 1 above)
Comment (philipp128) [reply skypy/galaxies/magnitude_error.py:96]: @rrjbca, what is the advantage using np.minimum?

File: skypy/galaxies/magnitude_error.py (hunk 2 above)
Comment (philipp128) [reply skypy/galaxies/magnitude_error.py:89]: It looks like there is a problem in the function when a scalar value for `zeropoint` is used. Shall I fix this here or in a separate PR?

**Review by rrjbca — CHANGES_REQUESTED** [review]

File: skypy/galaxies/magnitude_error.py (hunk 2 above)
Comment (rrjbca) [reply skypy/galaxies/magnitude_error.py]: No you just need to be careful with the minus sign convention used by the function. See my new review comment

File: skypy/galaxies/magnitude_error.py (hunk 1 above)
Comment (rrjbca) [reply skypy/galaxies/magnitude_error.py]: Less code, more readable, less maintenance, no branching, potentially better optimised by numpy...

File: skypy/utils/photometry.py [hunk 5]
```diff
@@ -13,6 +13,7 @@
     'luminosity_in_band',
     'mag_ab',
     'SpectrumTemplates',
+    'rykoff_error',
```
[the remark this answers is not in the document]
Comment (philipp128) [reply skypy/utils/photometry.py:16]: True. I agree with this

File: docs/galaxies.rst [hunk 6]
```diff
@@ -86,7 +86,7 @@ The following models are found in the `skypy.galaxies.stellar_mass` package.
 
 .. currentmodule:: skypy.galaxies.stellar_mass
 .. autosummary::
-  :nosignatures:
+   :nosignatures:
```
[the remark this answers is not in the document]
Comment (philipp128) [reply docs/galaxies.rst:89]: I don't get it. In my editor it looks correct. Maybe we can just change it here before merging if the other things are ok?

File: skypy/utils/photometry.py [hunk 7]
```diff
@@ -304,3 +305,90 @@ def absolute_magnitude_from_luminosity(luminosity, zeropoint=None):
...
+def magnitude_error_rykoff(magnitude, magnitude_limit, magnitude_zp, a, b, error_limit=np.inf):
+    r"""Magnitude error acoording to the model from Rykoff et al. (2015).
+
+    Given an apparent magnitude calculate the magnitude error that is introduced
+    by the survey specifications and follows the model described in Rykoff et al. (2015).
+
+    Parameters
+    ----------
+    magnitude: array_like
+        Galaxy apparent magnitude. This and the other array_like parameters must
```
Comment (rrjbca) [inline skypy/utils/photometry.py:319]:
```suggestion
        Apparent magnitude. This and the other array_like parameters must
```

File: skypy/utils/photometry.py [hunk 8]
```diff
@@ -304,3 +305,90 @@ def absolute_magnitude_from_luminosity(luminosity, zeropoint=None):
...
+        &= \frac{2.5}{\ln(10)} \left[ \frac{1}{Ft_{\mathrm{eff}}}
+            \left( 1 + \frac{F_{\mathrm{noise}}}{F} \right) \right]^{1/2} \;,
+
+    where
+
+    .. math::
+
+        F=10^{-0.4(m - m_{\mathrm{ZP}})}
+
+    is the galaxy's flux,
```
Comment (rrjbca) [inline skypy/utils/photometry.py:358]:
```suggestion
    is the source's flux,
```

File: docs/galaxies.rst (hunk 6 above)
Comment (rrjbca) [inline docs/galaxies.rst:89]:
```suggestion
  :nosignatures:
```

**Review by rrjbca — COMMENTED** [review]
Looks good. Really like the extra analytic case for the unit tests. Just spotted a couple of very small things I'll address myself quickly...

**Comment by philipp128** [comment]
> Looks good. Really like the extra analytic case for the unit tests. Just spotted a couple of very small things I'll address myself quickly...

Perfect. Thanks a lot :)

**Pull request merged**